src/components/connectAdvanced.js

```diff
@@ -100,203 +90,169 @@ export default function connectAdvanced(
-        // to any descendants receiving store+subscription from context; it passes along
-        // subscription passed to it. Otherwise, it shadows the parent subscription, which allows
-        // Connect to control ordering of notifications to flow top-down.
-        const subscription = this.propsMode ? null : this.subscription
-        return { [subscriptionKey]: subscription || this.context[subscriptionKey] }
-      }
-
-      componentDidMount() {
-        if (!shouldHandleStateChanges) return
-
-        // componentWillMount fires during server side rendering, but componentDidMount and
-        // componentWillUnmount do not. Because of this, trySubscribe happens during ...didMount.
-        // Otherwise, unsubscription would never take place during SSR, causing a memory leak.
-        // To handle the case where a child component may have triggered a state change by
-        // dispatching an action in its componentWillMount, we have to re-run the select and maybe
-        // re-render.
-        this.subscription.trySubscribe()
-        this.selector.run(this.props)
-        if (this.selector.shouldComponentUpdate) this.forceUpdate()
+        this.generatedDerivedProps = this.makeDerivedPropsGenerator()
+        this.renderWrappedComponent = this.renderWrappedComponent.bind(this)
       }
 
-      componentWillReceiveProps(nextProps) {
-        this.selector.run(nextProps)
-      }
-
-      shouldComponentUpdate() {
-        return this.selector.shouldComponentUpdate
-      }
-
-      componentWillUnmount() {
-        if (this.subscription) this.subscription.tryUnsubscribe()
-        this.subscription = null
-        this.notifyNestedSubs = noop
-        this.store = null
-        this.selector.run = noop
-        this.selector.shouldComponentUpdate = false
+      makeDerivedPropsGenerator() {
+        let lastProps
+        let lastState
+        let lastDerivedProps
+        let lastStore
+        let sourceSelector
+        return (state, props, store) => {
+          if ((connectOptions.pure && lastProps === props) && (lastState === state)) {
+            return lastDerivedProps
+          }
+          if (store !== lastStore) {
+            lastStore = store
+            sourceSelector = selectorFactory(store.dispatch, selectorFactoryOptions)
+          }
+          lastProps = props
+          lastState = state
+          const nextProps = sourceSelector(state, props)
+          if (lastDerivedProps === nextProps) {
+            return lastDerivedProps
+          }
+          lastDerivedProps = nextProps
+          return lastDerivedProps
+        }
       }
 
-      getWrappedInstance() {
-        invariant(withRef,
-          `To access the wrapped instance, you need to specify ` +
-          `{ withRef: true } in the options argument of the ${methodName}() call.`
+      renderWrappedComponentWithRef(value) {
+        invariant(value,
```

Comment:
Ditto for this.  I don't know how expensive the call to `invariant()` is, but I don't want to do this in prod.

Reply:
I thought the point of `invariant` is for it to always be used, even in prod. If it's something you can survive in prod it can be just a `warning`.

Reply:
@Kovensky is correct. invariant just ignores itself in production, so these are dev-only warnings

Reply:
Okay, good to know.